Debug output for transaction submit

newhoggy opened the pull request:
* A `scripts/plutus/example-txin-locking-plutus-script-fail.sh` to trigger the error message.
* New New transaction submit `--error-detail-out` option
* `ToJSON` instances for this feature
* Carefully annotate types so of values so it's easy to trace dependencies between `ToJSON` instances

newhoggy commented:
Sample run:

```
scripts/plutus/example-txin-locking-plutus-script-fail.sh
Socket path: example/node-bft1/node.sock
Socket path: /Users/jky/wrk/iohk/cardano-node
srwxr-xr-x  1 jky  staff  0  8 May 00:19 example/node-bft1/node.sock
Always succeeds Plutus script in use. Any datum and redeemer combination will succeed.
Script at: ./scripts/plutus/scripts/always-succeeds-spending.plutus
{
    "7ca831d2e870a434505d43e2d57f62a0dfc99a5868562ab1b8c88921933e3f25#0": {
        "address": "addr_test1vrpu544ueucqukajy09lr7g2ukre8y2l6xk0nw4qafxj29ggz6dk4",
        "datum": null,
        "datumhash": null,
        "value": {
            "lovelace": 900000000000
        }
    }
}Estimated transaction fee: Lovelace 169197
Transaction successfully submitted.
Pausing for 5 seconds...
{
    "a2d9e0c74e1e7e14d2a60cd1534897919b6edf77c8b89ed935eb1d801e9f380d#0": {
        "address": "addr_test1vrpu544ueucqukajy09lr7g2ukre8y2l6xk0nw4qafxj29ggz6dk4",
        "datum": null,
        "datumhash": null,
        "value": {
            "lovelace": 299999830803
        }
    },
    "a2d9e0c74e1e7e14d2a60cd1534897919b6edf77c8b89ed935eb1d801e9f380d#2": {
        "address": "addr_test1vrpu544ueucqukajy09lr7g2ukre8y2l6xk0nw4qafxj29ggz6dk4",
        "datum": null,
        "datumhash": null,
        "value": {
            "lovelace": 300000000000
        }
    }
}Plutus txin
a2d9e0c74e1e7e14d2a60cd1534897919b6edf77c8b89ed935eb1d801e9f380d#1|
Collateral
a2d9e0c74e1e7e14d2a60cd1534897919b6edf77c8b89ed935eb1d801e9f380d#0|
Submit the tx with plutus script and wait 5 seconds...
Command failed: transaction submit  Error: Error while submitting tx: ShelleyTxValidationError ShelleyBasedEraAlonzo (ApplyTxError [UtxowFailure (WrappedShelleyEraFailure (UtxoFailure (UtxosFailure (ValidationTagMismatch (IsValid True) (FailedUnexpectedly (PlutusFailure "\nThe 3 arg plutus script (PlutusScript PlutusV1 ScriptHash \"67f33146617a5e61936081db3b2117cbf59bd2123748f58ac9678656\") fails.\nCekError An error has occurred:  User error:\nThe budget was overspent. Final negative state: ({ cpu: -99 | mem: -99 })\nThe protocol version is: ProtVer {pvMajor = 2, pvMinor = 0}\nThe data is: I 42\nThe redeemer is: I 42\nThe context is:\nPurpose: Spending (TxOutRef {txOutRefId = a2d9e0c74e1e7e14d2a60cd1534897919b6edf77c8b89ed935eb1d801e9f380d, txOutRefIdx = 1})\nTxInfo:\n  TxId: b17390563d5a04c460aafa53d1308aaef544e16bdaf7a139c97b1b7501c7506b\n  Inputs: [ a2d9e0c74e1e7e14d2a60cd1534897919b6edf77c8b89ed935eb1d801e9f380d!1 -> - Value (Map [(,Map [(\"\",300000000000)])]) addressed to\n                                                                                    ScriptCredential: 67f33146617a5e61936081db3b2117cbf59bd2123748f58ac9678656 (no staking credential) ]\n  Outputs: [ - Value (Map [(,Map [(\"\",299980000000)])]) addressed to\n               PubKeyCredential: 4088059bbeb6add02eecd0c6a2a52c06910f2a6b4ba0029e9fe6ed13 (no staking credential) ]\n  Fee: Value (Map [(,Map [(\"\",20000000)])])\n  Value minted: Value (Map [])\n  DCerts: []\n  Wdrl: []\n  Valid range: (-\8734 , +\8734)\n  Signatories: []\n  Datums: [ ( 9e1199a988ba72ffd6e9c269cadb3b53b5f360ff99f112d9b2ee30c4d74ad88b\n          , 42 ) ]\n" "hgCYphoAAwJZAAEBGgAGC8cZAm0AARoAAknwGQPoAAEaAAJJ8BggGgAlzqgZcfcEGXRNGGQZdE0YZBl0TRhkGXRNGGQZdE0YZBl0TRhkGGQYZBl0TRhkGgACSfAYIBoAAknwGCAaAAJJ8BggGgACSfAZA+gAARoAAknwGCAaAAJJ8BkD6AAIGgACQiAaAAZ+Ixh2AAEBGgACSfAZA+gACBoAAknwGgABt5gY9wEaAAJJ8BknEAEaAAIVXhkFLgEZA+gaAAJJ8BkD6AEaAAJJ8BggGgACSfAYIBoAAknwGCABARoAAknwARoAAknwBBoAAZSvGPgBGgABlK8Y+AEaAAI3fBkFVgEaAAK96hkB8QEaAAJJ8BggGgACSfAYIBoAAknwGCAaAAJJ8BggGgACSfAYIBoAAknwGCAaAAJCIBoABn4jGHYAAQEZ8EwZK9IAARoAAknwGCAaAAJCIBoABn4jGHYAAQEaAAJCIBoABn4jGHYAAQEaACXOqBlx9wQAGgABQbsEGgACSfAZE4gAARoAAknwGCAaAAMCWQABARoAAknwGCAaAAJJ8BggGgACSfAYIBoAAknwGCAaAAJJ8BggGgACSfAYIBoAAknwGCAaADMNpwEBggEBTk0BAAAzIiIgBRIAEgARnxgqGCrYeZ/YeZ+f2Hmf2Hmf2HmfWCCi2eDHTh5+FNKmDNFTSJeRm27fd8i4ntk16x2AHp84Df8B/9h5n9h5n9h6n1gcZ/MxRmF6XmGTYIHbOyEXy/Wb0hI3SPWKyWeGVv/YeoD/oUChQBsAAABF2WS4ANh5n1ggnhGZqYi6cv/W6cJpyts7U7XzYP+Z8RLZsu4wxNdK2Iv/////n9h5n9h5n9h5n1gcQIgFm762rdAu7NDGoqUsBpEPKmtLoAKen+btE//YeoD/oUChQBsAAABF2DOLANh6gP//oUChQBoBMS0AoUChQACAgNh5n9h5n9h5gNh6gP/YeZ/Ye4DYeoD//4Cf2HmfWCCeEZmpiLpy/9bpwmnK2ztTtfNg/5nxEtmy7jDE10rYixgq///YeZ9YILFzkFY9WgTEYKr6U9Ewiq71ROFr2vehOcl7G3UBx1Br///Yep/YeZ/YeZ9YIKLZ4MdOHn4U0qYM0VNIl5Gbbt93yLie2TXrHYAenzgN/wH/////ggIA" :| []))))))])
```

newhoggy commented:
Sample JSON output:

[error-detail.json.gz](https://github.com/input-output-hk/cardano-node/files/8645350/error-detail.json.gz)

Jimbo4350 commented:
A few things:
- We shouldn't create another module for the JSON instances. We have `Cardano.Api.Orphans` for orphan instances.
- `Command failed: transaction submit  Error: Error while submitting tx: ShelleyTxValidationError ShelleyBasedEraAlonzo (ApplyTxError [UtxowFailure (WrappedShelleyEraFailure (UtxoFailure (UtxosFailure (ValidationTagMismatch (IsValid True) (FailedUnexpectedly (PlutusFailure "\nThe 3 arg plutus script (PlutusScript PlutusV1 ScriptHash...`  is a poorly formatted error message. It is poorly formatted because we are calling `show` in `renderShelleyTxCmdError`. We should also render the JSON here

```
    Alonzo.PlutusFailure t bs -> object
      [ "kind"                  .= String "FailureDescription"
      , "error"                 .= String "PlutusFailure"
      , "description"           .= t
      , "reconstructionDetail"  .= Alonzo.debugPlutus (BSU.toString bs)
      ]
```
We should not render `t` in the `description` field because can't (easily) control the formatting. We should only render `reconstructionDetail` as you have done. 

Lets start with the above and I'll have another look.

newhoggy commented:
I've started using the `prettyprint` library to output the error.

The output now looks like this.  It looks no so great, but with a bit of work, it could be improved a lot.

```
Command failed: transaction submit  Error: Error while submitting tx: ShelleyTxValidationError ShelleyBasedEraAlonzo:
kind: UtxowFailure
  value 1:
    [g1] WrappedShelleyEraFailure [ij] UtxosFailure [j1] ValidationTagMismatch:
    isValidating: Valid
      reason: FailedUnexpectedly
      [ PlutusFailure
      _
        DebugInfo [] PlutusErrorV1 An error has occurred:  User error:
        The budget was overspent. Final negative state: ({ cpu: -99
        | mem: -99
        }) PlutusDebugV1
        costModel: CostModel
          exUnits: ExUnits
          sbs: 4d01000033222220051200120011
          scriptHash: 67f33146617a5e61936081db3b2117cbf59bd2123748f58ac9678656
          ds: [ 42
              , 42
              , <<[<<<"/\228\236Sluh2#\134\212i\160?\209\246\199a\191\199\202\208\155Q\238\248\SUBC\172\151\206i">,
              1>,
              <<<"g\243\&1Faz^a\147`\129\219;!\ETB\203\245\155\210\DC27H\245\138\201g\134V">,
              <>>,
              {"": {"": 300000000000}},
              <"\158\DC1\153\169\136\186r\255\214\233\194i\202\219;S\181\243`\255\153\241\DC2\217\178\238\&0\196\215J\216\139">>>],
              [<<<"@\136\ENQ\155\190\182\173\208.\236\208\198\162\165,\ACK\145\SI*kK\160\STX\158\159\230\237\DC3">,
              <>>,
              {"": {"": 299980000000}},
              <>>],
              {"": {"": 20000000}},
              {"": {"": 0}},
              [],
              [],
              <<<>, <>>, <<>, <>>>,
              [],
              [<"\158\DC1\153\169\136\186r\255\214\233\194i\202\219;S\181\243`\255\153\241\DC2\217\178\238\&0\196\215J\216\139",
              42>],
              <"\186\210O\212\205>\199\211K\DC4\240|\206N\227,&\213\251\244f\233b\EM\142\149\176\203h\162\166\197">>,
              <<<"/\228\236Sluh2#\134\212i\160?\209\246\199a\191\199\202\208\155Q\238\248\SUBC\172\151\206i">,
              1>>> ]
          dsSummary: PlutusData { data = 42 , redeemer = 42 , info = Just ScriptContext TxInfo
            {
            , txInfoInputs = {"elements":[{"arguments":[{"arguments":[{"arguments":[{"kind":"B","value":"2fe4ec536c7568322386d469a03fd1f6c761bfc7cad09b51eef81a43ac97ce69"}],"kind":"Constr","tag":0},{"kind":"I","value":1}],"kind":"Constr","tag":0},{"arguments":[{"arguments":[{"arguments":[{"kind":"B","value":"67f33146617a5e61936081db3b2117cbf59bd2123748f58ac9678656"}],"kind":"Constr","tag":1},{"arguments":[],"kind":"Constr","tag":1}],"kind":"Constr","tag":0},{"entries":[[{"kind":"B","value":""},{"entries":[[{"kind":"B","value":""},{"kind":"I","value":300000000000}]],"kind":"Map"}]],"kind":"Map"},{"arguments":[{"kind":"B","value":"9e1199a988ba72ffd6e9c269cadb3b53b5f360ff99f112d9b2ee30c4d74ad88b"}],"kind":"Constr","tag":0}],"kind":"Constr","tag":0}],"kind":"Constr","tag":0}],"kind":"List"}
            , txInfoOutputs = {"elements":[{"arguments":[{"arguments":[{"arguments":[{"kind":"B","value":"4088059bbeb6add02eecd0c6a2a52c06910f2a6b4ba0029e9fe6ed13"}],"kind":"Constr","tag":0},{"arguments":[],"kind":"Constr","tag":1}],"kind":"Constr","tag":0},{"entries":[[{"kind":"B","value":""},{"entries":[[{"kind":"B","value":""},{"kind":"I","value":299980000000}]],"kind":"Map"}]],"kind":"Map"},{"arguments":[],"kind":"Constr","tag":1}],"kind":"Constr","tag":0}],"kind":"List"}
            , txInfoFee = {"entries":[[{"kind":"B","value":""},{"entries":[[{"kind":"B","value":""},{"kind":"I","value":20000000}]],"kind":"Map"}]],"kind":"Map"}
            , txInfoMint = {"entries":[[{"kind":"B","value":""},{"entries":[[{"kind":"B","value":""},{"kind":"I","value":0}]],"kind":"Map"}]],"kind":"Map"}
            , txInfoDCert = {"elements":[],"kind":"List"}
            , txInfoWdrl = {"elements":[],"kind":"List"}
            , txInfoValidRange = {"arguments":[{"arguments":[{"arguments":[],"kind":"Constr","tag":0},{"arguments":[],"kind":"Constr","tag":1}],"kind":"Constr","tag":0},{"arguments":[{"arguments":[],"kind":"Constr","tag":2},{"arguments":[],"kind":"Constr","tag":1}],"kind":"Constr","tag":0}],"kind":"Constr","tag":0}
            , txInfoSignatories = {"elements":[],"kind":"List"}
            , txInfoData = {"elements":[{"arguments":[{"kind":"B","value":"9e1199a988ba72ffd6e9c269cadb3b53b5f360ff99f112d9b2ee30c4d74ad88b"},{"kind":"I","value":42}],"kind":"Constr","tag":0}],"kind":"List"}
            , txInfoId = {"arguments":[{"kind":"B","value":"bad24fd4cd3ec7d34b14f07cce4ee32c26d5fbf466e962198e95b0cb68a2a6c5"}],"kind":"Constr","tag":0}
            } Spending <<"/\228\236Sluh2#\134\212i\160?\209\246\199a\191\199\202\208\155Q\238\248\SUBC\172\151\206i">,
            1> }
          protVer: ProtoVer {..} ]
```

newhoggy commented:
This is what it looks like now in YAML form.  I think it can be compressed further:

```
- kind: UtxowFailure
  value:
    error:
      isvalidating: true
      kind: ValidationTagMismatch
      reason:
        error: FailedUnexpectedly
        kind: TagMismatchDescription
        reconstruction:
          - error: PlutusFailure
            kind: FailureDescription
            reconstructionDetail:
              debug:
                dsSummary:
                  data:
                    kind: I
                    value: 42
                  info:
                    purpose:
                      kind: Spending
                      value:
                        arguments:
                          - arguments:
                              - kind: B
                                value: 1ec8d4197ec55c8b0ff17573f0b85d880466f0d7a40ff9b4e20e061cdce1f821
                            kind: Constr
                            tag: 0
                          - kind: I
                            value: 1
                        kind: Constr
                        tag: 0
                    txInfo:
                      dCert: []
                      data:
                        - - kind: B
                            value: 9e1199a988ba72ffd6e9c269cadb3b53b5f360ff99f112d9b2ee30c4d74ad88b
                          - kind: I
                            value: 42
                      fee:
                        entries:
                          - - kind: B
                              value: ""
                            - entries:
                                - - kind: B
                                    value: ""
                                  - kind: I
                                    value: 20000000
                              kind: Map
                        kind: Map
                      id: e234d3f3a0d613174459e1f213a437c7ed38bb6122fff32a90bfed18b691353d
                      inputs:
                        - arguments:
                            - arguments:
                                - arguments:
                                    - kind: B
                                      value: 1ec8d4197ec55c8b0ff17573f0b85d880466f0d7a40ff9b4e20e061cdce1f821
                                  kind: Constr
                                  tag: 0
                                - kind: I
                                  value: 1
                              kind: Constr
                              tag: 0
                            - arguments:
                                - arguments:
                                    - arguments:
                                        - kind: B
                                          value: 67f33146617a5e61936081db3b2117cbf59bd2123748f58ac9678656
                                      kind: Constr
                                      tag: 1
                                    - arguments: []
                                      kind: Constr
                                      tag: 1
                                  kind: Constr
                                  tag: 0
                                - entries:
                                    - - kind: B
                                        value: ""
                                      - entries:
                                          - - kind: B
                                              value: ""
                                            - kind: I
                                              value: 11111029646
                                        kind: Map
                                  kind: Map
                                - arguments:
                                    - kind: B
                                      value: 9e1199a988ba72ffd6e9c269cadb3b53b5f360ff99f112d9b2ee30c4d74ad88b
                                  kind: Constr
                                  tag: 0
                              kind: Constr
                              tag: 0
                          kind: Constr
                          tag: 0
                      mint: Value (Map [])
                      outputs:
                        - arguments:
                            - arguments:
                                - arguments:
                                    - kind: B
                                      value: 4088059bbeb6add02eecd0c6a2a52c06910f2a6b4ba0029e9fe6ed13
                                  kind: Constr
                                  tag: 0
                                - arguments: []
                                  kind: Constr
                                  tag: 1
                              kind: Constr
                              tag: 0
                            - entries:
                                - - kind: B
                                    value: ""
                                  - entries:
                                      - - kind: B
                                          value: ""
                                        - kind: I
                                          value: 299980000000
                                    kind: Map
                              kind: Map
                            - arguments: []
                              kind: Constr
                              tag: 1
                          kind: Constr
                          tag: 0
                      signatories: []
                      validRange:
                        arguments:
                          - arguments:
                              - arguments: []
                                kind: Constr
                                tag: 0
                              - arguments: []
                                kind: Constr
                                tag: 1
                            kind: Constr
                            tag: 0
                          - arguments:
                              - arguments: []
                                kind: Constr
                                tag: 2
                              - arguments: []
                                kind: Constr
                                tag: 1
                            kind: Constr
                            tag: 0
                        kind: Constr
                        tag: 0
                      wdrl: []
                  redeemer:
                    kind: I
                    value: 42
                exUnits:
                  exUnitsMem: 1
                  exUnitsSteps: 1
                protVer:
                  major: 2
                  minor: 0
                sbs: 4d01000033222220051200120011
                scriptHash: 67f33146617a5e61936081db3b2117cbf59bd2123748f58ac9678656
              error:
                error: |-
                  An error has occurred:  User error:
                  The budget was overspent. Final negative state: ({ cpu: -99 | mem: -99 })
                kind: CekError
                value:
                  _ewcCause: null
                  _ewcError:
                    UserEvaluationError:
                      budget:
                        exBudgetCPU: -99
                        exBudgetMemory: -99
                      kind: CekOutOfExError
              kind: DebugInfo
              texts: []
    kind: UtxosFailure
- kind: UtxowFailure
  value:
    consumed:
      lovelace: 11111029646
      policies: {}
    error: This transaction consumed Value 11111029646 (fromList []) but produced Value 300000000000 (fromList [])
    kind: ValueNotConservedUTxO
    produced:
      lovelace: 300000000000
      policies: {}
```

newhoggy commented:
Final version:

```
- kind: UtxowFailure
  value:
    error:
      isvalidating: true
      kind: ValidationTagMismatch
      reason:
        error: FailedUnexpectedly
        kind: TagMismatchDescription
        reconstruction:
          - error: PlutusFailure
            kind: FailureDescription
            reconstructionDetail:
              debug:
                dsSummary:
                  data: 42
                  info:
                    purpose:
                      kind: Spending
                      value:
                        - 07feafbc7e8e43210a16dab4ad341104d3e969804bc93a26b10010802dfab5a0
                        - 1
                    txInfo:
                      dCert: []
                      data:
                        - - 9e1199a988ba72ffd6e9c269cadb3b53b5f360ff99f112d9b2ee30c4d74ad88b
                          - 42
                      fee:
                        "":
                          "": 20000000
                      id: da282286e7ef57448277ee3e1232689e523c5b18096dd8862e042d97401a2d01
                      inputs:
                        - outRef:
                            - 07feafbc7e8e43210a16dab4ad341104d3e969804bc93a26b10010802dfab5a0
                            - 1
                          resolved:
                            address:
                              credential:
                                ScriptCredential: 67f33146617a5e61936081db3b2117cbf59bd2123748f58ac9678656
                              stakingCredential: null
                            datumHash: 9e1199a988ba72ffd6e9c269cadb3b53b5f360ff99f112d9b2ee30c4d74ad88b
                            value:
                              "":
                                "": 3703645216
                      mint:
                        "":
                          "": 0
                      outputs:
                        - address:
                            credential:
                              PubKeyCredential: 4088059bbeb6add02eecd0c6a2a52c06910f2a6b4ba0029e9fe6ed13
                            stakingCredential: null
                          datumHash: null
                          value:
                            "":
                              "": 299980000000
                      signatories: []
                      validRange:
                        - (
                        - -∞
                        - +∞
                        - )
                      wdrl: []
                  redeemer: 42
                exUnits:
                  exUnitsMem: 1
                  exUnitsSteps: 1
                protVer:
                  major: 2
                  minor: 0
                sbs: 4d01000033222220051200120011
                scriptHash: 67f33146617a5e61936081db3b2117cbf59bd2123748f58ac9678656
              error:
                error: |-
                  An error has occurred:  User error:
                  The budget was overspent. Final negative state: ({ cpu: -99 | mem: -99 })
                kind: CekError
                value:
                  _ewcCause: null
                  _ewcError:
                    UserEvaluationError:
                      budget:
                        exBudgetCPU: -99
                        exBudgetMemory: -99
                      kind: CekOutOfExError
              kind: DebugInfo
              texts: []
    kind: UtxosFailure
- kind: UtxowFailure
  value:
    consumed:
      lovelace: 3703645216
      policies: {}
    error: This transaction consumed Value 3703645216 (fromList []) but produced Value 300000000000 (fromList [])
    kind: ValueNotConservedUTxO
    produced:
      lovelace: 300000000000
      policies: {}
```

Jimbo4350 commented on cardano-api/src/Cardano/Api/Alonzo/Render.hs:
```diff
@@ -0,0 +1,73 @@
+ where
+  renderTuple :: (Alonzo.ScriptPurpose StandardCrypto, ScriptHash StandardCrypto) -> Pair
+  renderTuple (scriptPurpose, sHash) =
+    Aeson.fromText (renderScriptHash sHash) .= renderScriptPurpose scriptPurpose
```
So for example we don't need `renderScriptHash`:
```
-renderScriptHash :: ScriptHash StandardCrypto -> Text
-renderScriptHash = Api.serialiseToRawBytesHexText . Api.fromShelleyScriptHash
-
 renderMissingRedeemers :: [(Alonzo.ScriptPurpose StandardCrypto, ScriptHash StandardCrypto)] -> Value
 renderMissingRedeemers scripts = object $ map renderTuple  scripts
  where
   renderTuple :: (Alonzo.ScriptPurpose StandardCrypto, ScriptHash StandardCrypto) -> Pair
   renderTuple (scriptPurpose, sHash) =
-    Aeson.fromText (renderScriptHash sHash) .= renderScriptPurpose scriptPurpose
+    Aeson.fromText (Api.serialiseToRawBytesHexText $ Api.ScriptHash sHash) .= renderScriptPurpose scriptPurpose
 ```

Jimbo4350 replied:
We shouldn't need to add a `Cardano.Api.Alonzo.Render` module

newhoggy replied:
It's duplicate, but they are inaccessible because they are in `cardano-node`.

newhoggy replied:
The `Render` module is internal, so I don't think it will be exposed by the API.

newhoggy replied:
This module is gone.

Jimbo4350 reviewed (dismissed):
We can revisit this once more critical work has been completed. I will need to spend some more time carefully going through this PR because I think we can reduce the size of it significantly.

Jimbo4350 commented on cardano-api/src/Cardano/Api/Alonzo/Render.hs:
```diff
@@ -0,0 +1,73 @@
+module Cardano.Api.Alonzo.Render
```
A lot of this module duplicates functions that exist in cardano-node.

Jimbo4350 commented on cardano-api/src/Cardano/Api/InMode.hs:
```diff
@@ -1,7 +1,21 @@
-{-# LANGUAGE GADTs #-}
-{-# LANGUAGE ScopedTypeVariables #-}
-{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE DataKinds                  #-}
```
Why is your formatter aligning the pragmas?

newhoggy replied:
Fixed!

newhoggy commented:
I removed the `Render` module.  Of the render functions, I converted some to `ToJSON` instances where that made sense.  Others, I inlined, and the remainder, I put at the end of the `Orphans` module.

Jimbo4350 commented on cardano-api/src/Cardano/Api/Orphans.hs:
```diff
@@ -681,3 +778,1443 @@ instance Crypto.Crypto crypto => ToJSON (VMap VB VB (Shelley.KeyHash    'Shelley
+    , "failures"  .= do map toJSON fs :: [Value]
+    ]
+
+instance ToJSON ChainPredicateFailure where
```
So for example this `ToJSON` instance should not be necessary. Why? Because `ToObject ChainPredicateFailure` exists and the `ToObject` class has a `ToJSON` constraint: https://github.com/input-output-hk/iohk-monitoring-framework/blob/066f7002aac5a0efc20e49643fea45454f226caa/iohk-monitoring/src/Cardano/BM/Data/Tracer.lhs#L217
Can you double check that  the orphan instances you have added are needed?

Jimbo4350 commented on cardano-cli/src/Cardano/CLI/Shelley/Run/Transaction.hs:
```diff
@@ -1357,7 +1358,12 @@ runTxSubmit (AnyConsensusModeParams cModeParams) network txFile = do
+            let errorAsText = Text.pack $ show err
+            forM_ mErrorDetailFp $ \errorDetailFp ->
+              liftIO $ LBS.writeFile errorDetailFp (Aeson.encode err)
+            left $ ShelleyTxCmdTxSubmitError errorAsText
```
Should we also default the stdout to JSON? Ask QA what they think about this.

Jimbo4350 commented on scripts/plutus/example-txin-locking-plutus-script-fail.sh:
```diff
@@ -0,0 +1,136 @@
+#!/usr/bin/env bash
+
```
I'm not sure we want to add a new shell script just to see the failure.

Jimbo4350 reviewed (changes requested):
Couple of changes.
